Extended diagnostics integration

danieltre23 opened the pull request:
## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

Integrate extended template diagnostics with compiler and add `extendedTemplateDiagnostics` flag in the compiler options.

<!-- Please check the one that applies to this PR using "x". -->

- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [x] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:


## What is the current behavior?
The compiler currently can't access the `extended` package.

Issue Number: #42966 

## What is the new behavior?
Added a new phase in the compiler that runs all template checks and generates extended template diagnostics when the `extendedTemplateDiagnostics` and `strictTemplates` flags are set.

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

## Other information

danieltre23 committed:
refactor(compiler-cli): export `ExtendedTemplateCheckerImpl` and template checks

This commit exports the implementation of `ExtendedTemplateChecker` to
generate extended template diagnostics and all the template checks.

Refs #42966

danieltre23 committed:
refactor(compiler-cli): add `extendedTemplateCheck` phase to compiler

This commit integrates extended template checks with the compiler, by
adding another phase of diagnostics generation. This integration is
under the `_extendedTemplateDiagnostics` flag.

Refs #42966

danieltre23 committed:
test(compiler-cli): add extended template diagnostics end-to-end tests

This commit adds extended template diagnostics end-to-end tests, to make
sure the diagnostics are generated correctly. Template checks are
already tested with unit tests.

Refs #42966

dgp1130 commented:
I think you probably want to rebase this against `master` now that your previous PR has been merged. That should drop most of the extra commits here.

dgp1130 commented on packages/compiler-cli/src/ngtsc/core/src/compiler.ts:
```diff
@@ -894,6 +897,25 @@ export class NgCompiler {
     return this.nonTemplateDiagnostics;
   }
 
+  private getExtendedTemplateDiagnostics(sf?: ts.SourceFile): ts.Diagnostic[] {
```
A doc comment here explaining that not providing a source file returns extended diagnostics for *all* source files in the program would be helpful. It's a private function, so we don't need to be too detailed, but at least calling out what happens when not given a source file would be useful.

dgp1130 commented on packages/compiler-cli/src/ngtsc/transform/src/compilation.ts:
```diff
@@ -490,6 +490,29 @@ export class TraitCompiler implements ProgramTypeCheckAdapter {
+      return diagnostics;
+    }
+
+    for (const clazz of this.fileToClasses.get(sf)!) {
```
Nit: I generally recommend checking the result of `.get()` over using `.has()` and with a non-null assertion. Something like:

```typescript
const classes = this.fileToClasses.get(sf);
if (classes === undefined) {
  return [];
}

const diagnostics: ts.Diagnostic[] = [];
for (const clazz of classes) {
  // ...
}
```

This way we're able to leverage the type inference capabilities of the language and let the compiler validate that `classes` is defined where we expected it to be, rather than relying on a programmer-authored `!` operator.

dgp1130 commented on packages/compiler-cli/test/ngtsc/extended_template_diagnostics_spec.ts:
```diff
@@ -0,0 +1,103 @@
+
+      const diags = env.driveDiagnostics();
+      expect(diags.length).toBe(0);
+    });
```
Should we consider an error if `strictTemplates` is disabled but `extendedTemplateDiagnostics` are enabled? In that case, it's a configuration error and I think we should fail the compilation in this case and give a hard error to the user about it.

atscott replied:
Agreed, this should be an error.

dgp1130 reviewed (changes requested):
Looks good! Just a couple nits and a few questions for Andrew.

One other nit: The first commit message should probably have "refactor(compiler-cli): export `getExtendedTemplateDiagnosticsForComponent` and template checks" as its first line so it's more easily scanable when abbreviated to one line.

danieltre23 committed:
test(language-service): add extended diagnostics test

Add a test in the langauge-service to make sure the extended template
diagnostics are being correctly generated.

Refs #42966

atscott commented on packages/compiler-cli/src/ngtsc/core/api/src/public_options.ts:
```diff
@@ -252,6 +252,13 @@ export interface StrictTemplateOptions {
+   *
+   * Defaults to `false`, even if `fullTemplateTypeCheck` or `strictTemplates` are set.
+   */
+  extendedTemplateDiagnostics?: boolean;
```
`StrictTemplateOptions` isn't automatically extracted to `angular.io`. We document each option at https://github.com/angular/angular/blob/master/aio/content/guide/template-typecheck.md. So I think maybe just adding a sentence to the top to indicate it's for experimental, internal use only would work.

atscott commented on packages/compiler-cli/src/ngtsc/core/src/compiler.ts:
```diff
@@ -894,6 +897,25 @@ export class NgCompiler {
   }
 
+  private getExtendedTemplateDiagnostics(sf?: ts.SourceFile): ts.Diagnostic[] {
+    if (!this.options.extendedTemplateDiagnostics || !this.options.strictTemplates) {
```
Your understanding of `strictTemplates` is correct. Users _could_ enable it and also specifically disable each option. I don't think it's worth worrying about this case. Most devs won't be touching the options beyond what the cli does during the new project generation. Anyone who specifically edits those options should be aware of what that might mean for the compiler's understanding of the code.

dgp1130 replied:
SGTM. I think in that case extended checks wouldn't fail or cause any problems, they just wouldn't trigger some places where they otherwise would have. I think that's a reasonable degradation of behavior for having less accurate types.

atscott commented on packages/compiler-cli/src/ngtsc/transform/src/api.ts:
```diff
@@ -187,6 +187,10 @@ export interface DecoratorHandler<D, A, S extends SemanticSymbol|null, R> {
        resolution: Readonly<R>): void;
 
+  extendedTemplateCheck?
+      (component: ts.ClassDeclaration, tempalteTypeChecker: TemplateTypeChecker,
```
```suggestion
      (component: ts.ClassDeclaration, templateTypeChecker: TemplateTypeChecker,
```

atscott commented on packages/compiler-cli/test/ngtsc/extended_template_diagnostics_spec.ts:
```diff
@@ -0,0 +1,103 @@
+/**
```
nit: the commit header should use `test(compiler-cli)` rather than `refactor(compiler-cli)`

atscott commented on packages/compiler-cli/src/ngtsc/core/src/compiler.ts:
```diff
@@ -427,8 +427,10 @@ export class NgCompiler {
-        [...this.getNonTemplateDiagnostics(), ...this.getTemplateDiagnostics()]);
+    return this.addMessageTextDetails([
+      ...this.getNonTemplateDiagnostics(), ...this.getTemplateDiagnostics(),
+      ...this.getExtendedTemplateDiagnostics()
```
```suggestion
      ...this.getExtendedTemplateDiagnostics(),
```

atscott commented on packages/compiler-cli/src/ngtsc/core/src/compiler.ts:
```diff
@@ -439,7 +441,8 @@ export class NgCompiler {
       ...this.getNonTemplateDiagnostics().filter(diag => diag.file === file),
-      ...this.getTemplateDiagnosticsForFile(file, optimizeFor)
+      ...this.getTemplateDiagnosticsForFile(file, optimizeFor),
+      ...this.getExtendedTemplateDiagnostics(file)
```
```suggestion
      ...this.getExtendedTemplateDiagnostics(file),
```

atscott reviewed (changes requested):
Looks great overall! 

One thing about the `feat` commit in the middle: If we decide this is appropriately a `feat`, then we'll need to change the target to `minor`. If we want to merge to `patch`, that'll have to change to `refactor` because the tool won't allow `feat` commit in the patch branch.

dgp1130 commented on packages/compiler-cli/src/ngtsc/core/src/compiler.ts:
```diff
@@ -894,6 +898,34 @@ export class NgCompiler {
+  private getExtendedTemplateDiagnostics(sf?: ts.SourceFile): ts.Diagnostic[] {
+    if (!this.options.extendedTemplateDiagnostics || !this.options.strictTemplates) {
+      if (this.options.extendedTemplateDiagnostics) {
+        throw new Error(`The 'extendedTemplateDiagnostics' don't work without 'strictTemplates'.`);
```
Nit: phrasing.

"The 'extendedTemplateDiagnostics' option requires 'strictTemplates' to also be enabled."

dgp1130 commented on packages/compiler-cli/src/ngtsc/transform/src/compilation.ts:
```diff
@@ -490,6 +490,30 @@ export class TraitCompiler implements ProgramTypeCheckAdapter {
+        if (trait.handler.extendedTemplateCheck === undefined) {
+          continue;
+        }
+        if (isNamedClassDeclaration(clazz)) {
```
Nit: Since this is only dependent on `clazz`, could we move this up a bit to skip iterating over traits when not necessary? Something like:

```typescript
for (const clazz of classes) {
  if (!isNamedClassDeclaration(clazz)) continue;

  for (const trait of record.traits) {
    // ...
  }
}
```

dgp1130 reviewed (approved):
Looking good! Just a couple minor nits, but no major concerns here.

alxhub commented on packages/compiler-cli/src/ngtsc/core/src/compiler.ts:
```diff
@@ -425,8 +425,11 @@ export class NgCompiler {
+    return this.addMessageTextDetails([
+      ...this.getNonTemplateDiagnostics(),
+      ...this.getTemplateDiagnostics(),
+      ...this.getExtendedTemplateDiagnostics(),
```
Should we not check the flag here before adding extended diagnostics to the mix? I think this would be preferable to allocating an array for each call.

alxhub commented on packages/compiler-cli/src/ngtsc/core/src/compiler.ts:
```diff
@@ -892,6 +896,35 @@ export class NgCompiler {
+  private getExtendedTemplateDiagnostics(sf?: ts.SourceFile): ts.Diagnostic[] {
+    if (!this.options._extendedTemplateDiagnostics || !this.options.strictTemplates) {
+      if (this.options._extendedTemplateDiagnostics) {
+        throw new Error(
```
This check should probably happen at creation time, or as part of option diagnostics, not on-demand.

alxhub commented on packages/compiler-cli/src/ngtsc/annotations/src/component.ts:
```diff
@@ -612,6 +613,13 @@ export class ComponentDecoratorHandler implements
+      component: ts.ClassDeclaration, templateTypeChecker: TemplateTypeChecker,
+      typeChecker: ts.TypeChecker): ts.Diagnostic[] {
+    return getExtendedTemplateDiagnosticsForComponent(
+        component, templateTypeChecker, typeChecker, allTemplateChecks);
```
This is a little strange, architecturally.

1) the compiler (by that I mean the core `NgCompiler`) has no ability to control which template checks run. We're importing a random collection of checks from the _implementation_ of extended diagnostics, which this class should know nothing about.

2) we're receiving the `ts.TypeChecker` as an argument here, even though we already have a reference to it from our constructor

3) we're receiving the `TemplateTypeChecker` here, even though we're not going to do anything with it but pass it along.

Instead, we should architect `extendedTemplateCheck` similarly to how `typeCheck` works:

* Create an interface (`ExtendedTemplateChecker` maybe) which will represent the operation of performing extended template checks. `ComponentDecoratorHandler` can then depend on this interface without needing to care about how these checks work. Basically, this method becomes `return extendedTemplateChecker.check(component)`.

* We can then have an implementation `ExtendedTemplateCheckerImpl` which captures any needed dependencies (`ts.TypeChecker`, `TemplateTypeChecker`, and any requested `TemplateCheck`s in its `constructor`.

* `ExtendedTemplateCheckerImpl` can be constructed by the `NgCompiler` (with some number of `TemplateCheck`s selected) and then `extendedTemplateCheck` can be called with that checker.

Using the interface means that `ComponentDecoratorHandler` and `TraitCompiler` don't need to depend on the implementation, nor do they need to be concerned with what checks are being run or what services are required to implement them.

danieltre23 committed:
refactor(compiler-cli): add `ExtendedTemplateChecker`

Change the current way to run template checks to the
`ExtendedTemplateChecker` instead of just the
`getExtendedTemplateDiagnosticsForComponent` function. Refactored the
tests that used the previous function to use the new class.

Refs #42966

alxhub commented on packages/compiler-cli/src/ngtsc/core/src/compiler.ts:
```diff
@@ -316,6 +317,12 @@ export class NgCompiler {
   ) {
+    if (this.options._extendedTemplateDiagnostics === true &&
+        this.options.strictTemplates === false) {
+      throw new Error(
```
Consider moving this check into `verifyCompatibleTypeCheckOptions`.

danieltre23 replied:
Talked offline and concluded that moving the check into `verifyCompatibleTypeCheckOptions` would require a new `ErrorCode` for the diagnostic, and since the `_extendedTemplateDiagnostics` flag is for internal use only it wouldn't make sense to have a public api `ErrorCode` for that.

alxhub commented on packages/compiler-cli/src/ngtsc/core/src/compiler.ts:
```diff
@@ -425,8 +432,17 @@ export class NgCompiler {
   getDiagnostics(): ts.Diagnostic[] {
-    return this.addMessageTextDetails(
-        [...this.getNonTemplateDiagnostics(), ...this.getTemplateDiagnostics()]);
+    if (this.options._extendedTemplateDiagnostics) {
```
I think it'd be cleaner to implement this by constructing the main diagnostics array and then `.push()`ing the extended diagnostics if requested.

alxhub commented on packages/compiler-cli/src/ngtsc/core/src/compiler.ts:
```diff
@@ -892,6 +915,29 @@ export class NgCompiler {
+    const compilation = this.ensureAnalyzed();
+    const typeChecker = this.inputProgram.getTypeChecker();
+    const extendedTemplateChecker = new ExtendedTemplateCheckerImpl(
+        compilation.templateTypeChecker, typeChecker, allTemplateChecks);
```
Instead of having `allTemplateChecks` imported as a constant, we should explicitly list out and construct the checks here.

alxhub commented on packages/compiler-cli/src/ngtsc/transform/src/compilation.ts:
```diff
@@ -490,6 +491,28 @@ export class TraitCompiler implements ProgramTypeCheckAdapter {
+
+    const diagnostics: ts.Diagnostic[] = [];
+    for (const clazz of classes) {
+      if (isNamedClassDeclaration(clazz)) {
```
Invert this condition:
```suggestion
      if (!isNamedClassDeclaration(clazz)) {
        continue;
      }
```

This keeps the rest of the code cleaner - the "main case" should always be the least indented and guarded.

alxhub commented on packages/compiler-cli/src/ngtsc/typecheck/extended/index.ts:
```diff
@@ -0,0 +1,15 @@
+ */
+
+import {ErrorCode} from '../../diagnostics';
+import {TemplateCheck} from './api/api';
```
When you have a separate package, always import from the index:
```suggestion
import {TemplateCheck} from './api';
```

alxhub commented on packages/compiler-cli/src/ngtsc/typecheck/extended/index.ts:
```diff
@@ -0,0 +1,15 @@
+
+export {ExtendedTemplateCheckerImpl} from './src/extended_template_checker';
+
+export const allTemplateChecks: TemplateCheck<ErrorCode>[] = [new InvalidBananaInBoxCheck()];
```
1) don't declare things in the `index.ts` file, that's exclusively for re-exports
2) this doesn't need to be a constant anyway, it should be the consumer's responsibility to select which checks they want to use.

alxhub commented on packages/compiler-cli/src/ngtsc/typecheck/extended/BUILD.bazel:
```diff
@@ -10,6 +10,7 @@ ts_library(
         "//packages/compiler-cli/src/ngtsc/diagnostics",
         "//packages/compiler-cli/src/ngtsc/typecheck/api",
         "//packages/compiler-cli/src/ngtsc/typecheck/extended/api",
+        "//packages/compiler-cli/src/ngtsc/typecheck/extended/src/template_checks/invalid_banana_in_box",
```
This shouldn't be needed - only consumers should depend on the specific checks that they want to use.
